File: feature-ride/src/main/java/com/levnovikov/feature_ride/ride/BehaviorField.java
package com.levnovikov.feature_ride.ride;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

/**
 * Created by lev.novikov
 * Date: 19/12/17.
 */

public abstract class BehaviorField<T> {

    private BehaviorSubject<T> field = BehaviorSubject.create();
    public Observable<T> getStream() {
        return field.distinctUntilChanged();
    }

    public void set(T entity) {
        save(entity);
        field.onNext(entity);
    }

    abstract void save(T entity);
}

File: feature-ride/src/main/java/com/levnovikov/feature_ride/ride/BehaviorField.kt
package com.levnovikov.feature_ride.ride

import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

/**
 * Created by lev.novikov
 * Date: 19/12/17.
 */

abstract class BehaviorField<T> {

    private val field = BehaviorSubject.create<T>()
    fun stream(): Observable<T> = field.distinctUntilChanged()

    fun set(entity: T) {
        save(entity)
        field.onNext(entity)
    }

    internal abstract fun save(entity: T)
}
